src/test/matchers/matchers.ts

```diff
@@ -14,198 +14,206 @@
+  }, expected, { ...options, matchSubstring: true });
 }
 
 export function toHaveAttribute(
   this: ReturnType<Expect['getState']>,
-  locator: Locator,
+  locator: LocatorEx,
   name: string,
   expected: string | RegExp,
   options?: { timeout?: number },
 ) {
-  return toMatchText.call(this, 'toHaveAttribute', locator, 'Locator', async (expected, isNot, timeout) => {
-    return await (locator as any)._expect('to.have.attribute', { expected, isNot, timeout, data: { name } });
+  return toMatchText.call(this, 'toHaveAttribute', locator, 'Locator', async (isNot, timeout) => {
+    const expectedText = toExpectedTextValues([expected]);
+    return await locator._expect('to.have.attribute', { expressionArg: name, expectedText, isNot, timeout });
   }, expected, options);
 }
 
 export function toHaveClass(
   this: ReturnType<Expect['getState']>,
-  locator: Locator,
+  locator: LocatorEx,
   expected: string | RegExp | (string | RegExp)[],
   options?: { timeout?: number },
 ) {
   if (Array.isArray(expected)) {
-    return toEqual.call(this, 'toHaveClass', locator, 'Locator', async () => {
-      return await locator.evaluateAll(ee => ee.map(e => e.className));
+    return toEqual.call(this, 'toHaveClass', locator, 'Locator', async (isNot, timeout) => {
+      const expectedText = toExpectedTextValues(expected);
+      return await locator._expect('to.have.class.array', { expectedText, isNot, timeout });
     }, expected, options);
   } else {
-    return toMatchText.call(this, 'toHaveClass', locator, 'Locator', async (expected, isNot, timeout) => {
-      return await (locator as any)._expect('to.have.class', { expected, isNot, timeout });
+    return toMatchText.call(this, 'toHaveClass', locator, 'Locator', async (isNot, timeout) => {
+      const expectedText = toExpectedTextValues([expected]);
+      return await locator._expect('to.have.class', { expectedText, isNot, timeout });
     }, expected, options);
   }
 }
 
 export function toHaveCount(
   this: ReturnType<Expect['getState']>,
-  locator: Locator,
+  locator: LocatorEx,
   expected: number,
   options?: { timeout?: number },
 ) {
-  return toEqual.call(this, 'toHaveCount', locator, 'Locator', async timeout => {
-    return await locator.count();
+  return toEqual.call(this, 'toHaveCount', locator, 'Locator', async (isNot, timeout) => {
+    return await locator._expect('to.have.count', { expectedNumber: expected, isNot, timeout });
   }, expected, options);
 }
 
 export function toHaveCSS(
   this: ReturnType<Expect['getState']>,
-  locator: Locator,
+  locator: LocatorEx,
   name: string,
   expected: string | RegExp,
   options?: { timeout?: number },
 ) {
-  return toMatchText.call(this, 'toHaveCSS', locator, 'Locator', async (expected, isNot, timeout) => {
-    return await (locator as any)._expect('to.have.css', { expected, isNot, timeout, data: { name } });
+  return toMatchText.call(this, 'toHaveCSS', locator, 'Locator', async (isNot, timeout) => {
```

Comment:
(for the follow up): I guess we do not have to pass `timeout`/`options` through `toMatchText` anymore, because timeout does not change.